transformer_engine/jax/csrc/modules.cpp

```diff
@@ -300,59 +311,90 @@ void LayerNormForwardImpl(size_t n, size_t hidden, bool zero_centered_gamma, flo
 
-    auto barrier_tensor =
-        TensorWrapper(reinterpret_cast<char *>(workspace) + dummy_workspace_tensor.shape().data[0],
-                      dummy_barrier_tensor.shape(), dummy_barrier_tensor.dtype());
+    // dummy tensor wrappers that will carry workspace size info later
+    TensorWrapper dummy_work_tensor, dummy_barrier_tensor;
+    TensorWrapper dummy_dgamma_part_tensor, dummy_dbeta_part_tensor;
+    auto num_sm = cudaDevicePropertiesManager::Instance().GetMultiProcessorCount();
+    auto layernorm_bwd_func = zero_centered_gamma ? nvte_layernorm1p_bwd : nvte_layernorm_bwd;
 
+    // initialize dBeta information here -- layernorm will modify but RMSnorm will not
+    std::vector<size_t> dbeta_part_shape;
     if (is_layer_norm) {
-        auto beta_tensor = TensorWrapper(bias, weight_shape, w_dtype);
-        auto mu_tensor = TensorWrapper(mu, intermediates_shape, DType::kFloat32);
+        auto mu_tensor = TensorWrapper(nullptr, intermediates_shape, intermediates_dtype);
+        auto dbeta_tensor = TensorWrapper(nullptr, weight_shape, w_dtype);
 
-        layernorm_fwd_func(input_tensor.data(), gamma_tensor.data(), beta_tensor.data(), eps,
-                           output_tensor.data(), mu_tensor.data(), rsigma_tensor.data(), stream,
-                           num_sm, workspace_tensor.data(), barrier_tensor.data());
+        layernorm_bwd_func(dz_tensor.data(), x_tensor.data(), mu_tensor.data(),
+                           rsigma_tensor.data(), gamma_tensor.data(), xgrad_tensor.data(),
+                           wgrad_tensor.data(), dbeta_tensor.data(),
+                           dummy_dgamma_part_tensor.data(), dummy_dbeta_part_tensor.data(), nullptr,
+                           num_sm, dummy_work_tensor.data(), dummy_barrier_tensor.data());
+
+        dbeta_part_shape = MakeShapeVector(dummy_dbeta_part_tensor.shape());
     } else {
-        nvte_rmsnorm_fwd(input_tensor.data(), gamma_tensor.data(), eps, output_tensor.data(),
-                         rsigma_tensor.data(), stream, num_sm, workspace_tensor.data(),
-                         barrier_tensor.data());
+        NVTE_CHECK(!zero_centered_gamma, "rmsnorm doesn't support zero_centered_gamma.");
+        nvte_rmsnorm_bwd(dz_tensor.data(), x_tensor.data(), rsigma_tensor.data(),
+                         gamma_tensor.data(), xgrad_tensor.data(), wgrad_tensor.data(),
+                         dummy_dgamma_part_tensor.data(), nullptr,
+                         num_sm, dummy_work_tensor.data(), dummy_barrier_tensor.data());
+
+        dbeta_part_shape = std::vector<size_t>{0, 0};
     }
+
+    std::vector<size_t> work_shape = MakeShapeVector(dummy_work_tensor.shape());
+    std::vector<size_t> barrier_shape = MakeShapeVector(dummy_barrier_tensor.shape());
+    std::vector<size_t> dgamma_part_shape = MakeShapeVector(dummy_dgamma_part_tensor.shape());
+    return pybind11::make_tuple(std::make_pair(work_shape, dummy_work_tensor.dtype()),
```

Comment:
It seems all workspace help functions return a similar structure, like
`py11_tuple(pair(shape, dtype), ...). How about to design a util function, like `pack_workspace_return`?

Reply:
I did this at first but it seemed uglier to me than tuple(pair(shape, dtype),...) approach.

The util function would need to handle variable numbers of (shape, dtype) pairs. That means we either do preprocessor macro hacks or variadic templates, neither of which are readable or easy to maintain.

The explicit tuple(pair(...), ...) style also has the benefit of making the pair relationships very obvious at a quick glance, which I found helpful in debugging.